### ISSUE TITLE
unexpected input char in normal mode

### ISSUE THREAD

**Issue by zhang-stephen** [vim/vim#7844]
**Describe the bug**
pressing `j` twice quickly might be recognized as `ESC` twice.

**To Reproduce**
Detailed steps to reproduce the behavior:
1. use vim-plug to install [coc.nvim](https://github.com/neoclide/coc.nvim) and [coc-clangd](https://github.com/clangd/coc-clangd), and using package manager, like apt or pacman,  to install clangd
2. using the following as your vimrc
  ```
  set nocompatible
set showcmd
  call plug#begin('~/.vim/plugged')
Plug 'neoclide/coc.nvim', { 'branch': 'release' }
call plug#end()
  ```
3. open a c source file and run `: call CocActionAsync("jumpReferences")` on a symbol, to open view window of this symbol
4. using `j` to navigate and view references, and press quickly, you would see `^[^[`(2 ESCs) and the reference window closed

the `^[^[` is not expected, but it appeared and leaded to mistaking behaviors.

**Screenshots**
None

**Environment (please complete the following information):**
 - Vim version: 8.2.2849 and 8.2.2516
 - OS: Linux Manjaro 21-pre1, kernel 5.16.1(latest unstable)
 - Terminal: bash and tmux

**Additional context**
I don't think this is error from plugins. I have tried add code to print the input from `safe_vgetc()` and the output of option `showcmd`, `showcmd_buf[]` in `src/normal.c`. And I found they are correct, so could you tell me what happened? and how to fix it?

**Comment by lacygoill** [comment]
> Vim version: 8.2.2849 and 8.2.2516

There is no 8.2.2849 patch.

> Terminal: bash and tmux

Try to reproduce outside tmux.

> I don't think this is error from plugins.

We can't be sure of that until you provide a [minimal working example](https://en.wikipedia.org/wiki/Minimal_working_example).  See [here](https://github.com/chrisbra/vim_faq/blob/7f9c65c91d01e289a92b9e43a70ee2becadd66ce/vim_faq.txt#L1153-L1228) for how to get one.  Note that what is written there is not limited to the vimrc; you can use the same kind of technique to third-party plugins.  But the bigger and more complex a plugin is, the more time-consuming the task is.

Check out the bug tracker of the plugin; maybe someone already reported the issue.  Otherwise, open a new issue there.

**Comment by zhang-stephen** [comment]
Sorry, my faults....

I tested this with patch `8.2.2489` and `8.2.2516` on Linux Manjaro 21-pre1, outside the Tmux, and it reproduced.

I tested this with another instance of vim, patch `8.2.22**`(I cannot remember the patch number) on CentOS 7, and it does not reproduced.

I think what I gave is minimal working sample because I cannot reproduce this problem outside the `coc-list-location` window. But log from vim told me the key I pressed is `ESC` when I pressed `j`. I am confused about this.

My working environment is very simple: There is only a bash except Tmux, no GUI or others

**Comment by brammool** [comment]
This sounds very much like a problem with those coc plugins.  Please find out where to report problems for them.
If you think it's not caused by a plugin, then please find reproduction steps without those plugins.

**Comment by lacygoill** [comment]
> I tested this with patch 8.2.2489 and 8.2.2516 on Linux Manjaro 21-pre1, outside the Tmux, and it reproduced.

Thank you for checking it does reproduce outside tmux.  Try to remove tmux out of the equation.  The latter adds a layer of complexity which makes it much harder to debug Vim issues.

> I tested this with another instance of vim, patch 8.2.22**(I cannot remember the patch number) on CentOS 7, and it does not reproduced.

So, IIUC, it reproduces in some environments but not all.  That's interesting.  Maybe the issue depends on the terminal.  When you can reproduce, what is the real name of the terminal emulator you're using, and what's the output of the shell command `echo $TERM`?  Same questions when you cannot reproduce.  Can you see some pattern?   Like maybe `$TERM` has the same value in the terminals where you can reproduce the issue?

> I think what I gave is minimal working sample because I cannot reproduce this problem outside the coc-list-location window.

This is what [`cloc(1)`](https://github.com/AlDanial/cloc) tells me about the `coc.nvim` plugin:

    -------------------------------------------------------------------------------
    Language                     files          blank        comment           code
    -------------------------------------------------------------------------------
    vim script                      17            428            266           5470
    JSON                             2              1              0           1273
    JavaScript                       2             23            105            275
    YAML                             2              6              1             53
    Bourne Shell                     1              2              0             10
    -------------------------------------------------------------------------------
    SUM:                            24            460            372           7081
    -------------------------------------------------------------------------------

If we need 7k lines of code to reproduce a bug, it's definitely not minimal.

If there is a bug in Vim, and if you want it to be fixed as soon as possible, finding a minimal example which doesn't depend on plugins helps a lot.

> But log from vim told me the key I pressed is ESC when I pressed j. I am confused about this.

Maybe the plugin does something unexpected, like remapping the keys, or set some terminal options which it shouldn't...

**Comment by zhang-stephen** [comment]
I had compiled `patch-2489` and `patch-2516` on that CentOS 7, and it reproduced. Their `$TERM`  is same as each other, the value is `xterm-256color`.

>Maybe the plugin does something unexpected, like remapping the keys, or set some terminal options which it shouldn't...

>This sounds very much like a problem with those coc plugins. Please find out where to report problems for them.
If you think it's not caused by a plugin, then please find reproduction steps without those plugins.

Now I cannot reproduce it without coc.nvim and its plugin... But I think the keys are read by vim and forwarded them  to coc.nvim aren't them? In the vim source files, where to find the original key value, where to find the mapping opeartions, and where to find the communication with plugins? 

In short, where to find the code documentation for vim developing? I want to print some log to verify this porblem.

**Comment by chrisbra** [comment]
Try `vim -u NONE -N --cmd "call ch_logfile('cygwin.log', 'w')"` This should print all key codes Vim receives.

**Comment by lacygoill** [comment]
> I want to print some log to verify this porblem.

> Try vim -u NONE -N --cmd "call ch_logfile('cygwin.log', 'w')" This should print all key codes Vim receives.

Any update?

**Comment by zhang-stephen** [comment]
I think it was a env problem, because I could reproduce it on a linux manjaro VM only, with same version of vim.

**Comment by zhang-stephen** [comment]
so, close it now
